Add full-date compliant date format validation

sasamuku opened the pull request:
resolve https://github.com/ota42y/openapi_parser/issues/163

ota42y commented:
ISO8601 defines the following two formats as valid formats ( and ruby can parse both )
https://en.wikipedia.org/wiki/ISO_8601#Calendar_dates
```
> Date.iso8601("2016-04-01") # Extended format
=> #<Date: 2016-04-01 ((2457480j,0s,0n),+0s,2299161j)>
> Date.iso8601("20160401") # Basic format
=> #<Date: 2016-04-01 ((2457480j,0s,0n),+0s,2299161j)>
```


`format: date` is defined as `full-date` in RFC3339 for both OAS 3.0.3 and 3.1.0.
- 3.0.3
	- https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#data-types
- 3.1.0
	- We can use format defined by [JSON Schema Validation vocabulary](https://tools.ietf.org/html/draft-bhutton-json-schema-validation-00#section-7.3)
		- https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#data-types
	- `format: date` is defined as `full-date` in RFC3339 
		- https://datatracker.ietf.org/doc/html/draft-bhutton-json-schema-validation-00#section-7.3

`full-date` in RFC3339 is defined by as 
```
   date-fullyear   = 4DIGIT
   date-month      = 2DIGIT  ; 01-12
   date-mday       = 2DIGIT  ; 01-28, 01-29, 01-30, 01-31 based on
                             ; month/year
   full-date       = date-fullyear "-" date-month "-" date-mday
```

`full-date` in RFC3339 is defined as  `YYYY-MM-DD` only.

In ISO 8601, this format is called an extended format, and in addition to this format, basic format (`YYYYMMDD`) is also defined as valid.
Ruby's Date#iso8601 is processed as valid in both the basic and extended formats.

Also, the old definition `ISO 8601:2000` allows expressions such as `YYMMDD`, and this method in ruby ​​also treats this expression as valid (this format removed in `ISO 8601:2004`).
```
> Date.iso8601("160401").iso8601 # YYMMDD is valid format
=> "2016-04-01"
```

For these reasons, I think that ruby's Date#iso8601 cannot be used because it proceed invalid request in RFC3339 as valid input.

sasamuku commented:
@ota42y 
Thank you for your review!

Your concern is entirely correct.
The title of this PR is wrong. What I really want to do is not to comply with ISO 8601 but `full-date` format in RFC 3339. The implementation is now compliant with `full-date` not ISO 8601.

As you said, ISO 8601 allows both basic format and extended format, but RFC 3339 allows extended fromat (`full-date`) only. Additinaly, `format: date` in OAS should accept only `full-date`.

That's why I added the following two-stage checks to reject the basic format.
https://github.com/ota42y/openapi_parser/blob/a6df785332cf27e3755d2f43d80a4ce709915ae1/lib/openapi_parser/schema_validator/string_validator.rb#L77-L85

`Date#iso8601` returns string formated as `full-date`, such as  `2001-02-03`.
https://github.com/ruby/date/blob/435dfec6c8e917f77f4170a9232438ad9c0ace20/ext/date/date_core.c#L4607

Consider the following pattern as input:

```ruby
> Date.iso8601("21-02-12").iso8601 == "21-02-12"
=> false
> Date.iso8601("20160401").iso8601 == "20160401"
=> false
> Date.iso8601("160401").iso8601 == "160401"
=> false
> Date.iso8601("2021-02-12T12:59:00.000+09:00").iso8601 == "2021-02-12T12:59:00.000+09:00"
=> false
> Date.iso8601("2021-02-12").iso8601 == "2021-02-12"
=> true
```

The value compliant with `full-date` can only return `true`.
The implementation supports this behavior.

If you agree, I would remake the issue and PR with the correct title.

ota42y commented:
`parsed_date.iso8601 == value `
I see, this line verifies that the input is in RFC3339 format.
My understanding was unclear, so please change the title of the PR, this code is perfect!

I think you don't need remake PR... 
Wasn't it a github specification that the PR title cannot be changed if you fork?

sasamuku commented:
@ota42y 
As you said, `parsed_date.iso8601 == value` may lead to wrong understanding.
I modified it as follows:
`parsed_date.strftime('%Y-%m-%d') == value`

I also changed the title of the PR!

ota42y commented:
Note: sasamuku became a collaborator 🎉
